src/common/regions.ts

```diff
@@ -0,0 +1,146 @@
+import { chain, fromPairs } from 'lodash';
+import urlJoin from 'url-join';
+import US_STATE_DATASET from 'components/MapSelectors/datasets/us_states_dataset_01_02_2020.json';
+import countyAdjacencyMsa from './data/county_adjacency_msa.json';
+const { state_dataset, state_county_map_dataset } = US_STATE_DATASET;
+
+export type FipsCode = string;
+
+export enum RegionType {
+  COUNTY = 'County',
+  STATE = 'State',
+  MSA = 'MSA',
+}
+
+abstract class Region {
+  constructor(
+    public name: string,
+    public urlName: string,
+    public fipsCode: FipsCode,
+    public population: number,
+    public regionType: RegionType,
+  ) {}
+
+  abstract fullName(): string;
+  abstract relativeUrl(): string;
+  canonicalUrl() {
+    return urlJoin('https://covidactnow.org', this.relativeUrl());
+  }
+}
+
+export class State extends Region {
+  constructor(
+    public name: string,
+    public urlName: string,
+    public fipsCode: FipsCode,
+    public population: number,
+    public stateCode: string,
+  ) {
+    super(name, urlName, fipsCode, population, RegionType.STATE);
+  }
+
+  fullName() {
+    return this.name;
+  }
+
+  relativeUrl() {
+    return `us/${this.urlName}`;
+  }
+}
+
+export class County extends Region {
+  constructor(
+    public name: string,
+    public urlName: string,
+    public fipsCode: FipsCode,
+    public population: number,
+    public state: State,
+    public cities: string[],
+    private adjacentCountiesFips: FipsCode[],
+  ) {
+    super(name, urlName, fipsCode, population, RegionType.COUNTY);
+  }
+
+  fullName() {
+    return `${this.name}, ${this.state.stateCode}`;
+  }
+
+  relativeUrl() {
+    return urlJoin(this.state.relativeUrl(), `county/${this.urlName}`);
+  }
+}
+
+const states: State[] = chain(state_dataset)
+  .map(stateInfo => {
+    return new State(
+      stateInfo.state,
+      stateInfo.state_url_name,
+      stateInfo.state_fips_code,
+      stateInfo.population,
+      stateInfo.state_code,
+    );
+  })
+  .value();
+
+// TODO: Aggregated regions
+
+const statesByFips = fromPairs(states.map(state => [state.fipsCode, state]));
+
+interface AdjacencyInfo {
+  adjacent_counties: FipsCode[];
+  msa_code?: string;
+}
+
+const adjacency: { [fipsCode: string]: AdjacencyInfo } =
+  countyAdjacencyMsa.counties;
+
+const counties: County[] = chain(state_county_map_dataset)
+  .map(stateData => stateData.county_dataset)
+  .flatten()
+  .filter(county => Object.keys(statesByFips).includes(county.state_fips_code))
+  .map(countyInfo => {
+    const countyFips = countyInfo.full_fips_code;
+    const state = statesByFips[countyInfo.state_fips_code];
+    const adjacentCounties = adjacency[countyFips]?.adjacent_counties;
+    return new County(
+      countyInfo.county,
+      countyInfo.county_url_name,
+      countyInfo.full_fips_code,
+      countyInfo.population,
+      state,
+      countyInfo.cities || [],
+      adjacentCounties || [],
+    );
+  })
+  .value();
+
+class RegionDB {
```

Comment:
I'm not 100% sure, but using a class like this feels like it might not be as flexible as using a hook to filter by different regions.  

My thinking is that there are cases where we'd want to work with subsets of regions, and having selectors that will update if the subset of regions changes would be useful. 

I guess if this is the global list of regions, that should be fine, but it might be something to think about for different access patterns.

Reply:
I reviewed the access patterns, these are the ones I could find:
- all counties
- all states
- all counties in a state
- all metro counties in a state
- all non-metro counties in a state
- all counties adjacent to a given county

let me know if you find missing queries. I think that this is enough for now, as we can always filter directly using `.all()` or with the convenience methods by region type

Reply:
I'm not 100% sure what "using a hook to filter by different regions" refers to.  If we're talking about a react hook, that feels like it could easily built on top of this, so this still feels like a good primitive.